sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StatePartitionReader.scala

```diff
@@ -166,16 +184,22 @@ class StatePartitionReader(
         stateVariableInfoOpt match {
           case Some(stateVarInfo) =>
             val stateVarType = stateVarInfo.stateVariableType
-            val hasTTLEnabled = stateVarInfo.ttlEnabled
 
             stateVarType match {
               case StateVariableType.ValueState =>
-                if (hasTTLEnabled) {
-                  SchemaUtil.unifyStateRowPairWithTTL((pair.key, pair.value), valueSchema,
-                    partition.partition)
-                } else {
-                  SchemaUtil.unifyStateRowPair((pair.key, pair.value), partition.partition)
+                SchemaUtil.unifyStateRowPair((pair.key, pair.value), partition.partition)
+
+              case StateVariableType.ListState =>
+                val key = pair.key
+                val result = store.valuesIterator(key, stateVarName)
+                var unsafeRowArr: Seq[UnsafeRow] = Seq.empty
```

Comment:
Worth noting that we are materializing every element in the list, so despite we avoid memory issue in ListState, the problem will pop up when reading ListState via state data source reader.

I guess it's probably worth reconsidering the schema and UX. e.g. having `index` and `value` as columns for the value of list state and allow multiple rows for the same state value. (You've used explode in the example but we could just provide that result directly.) You'll also have the same issue with MapType as well, so worth considering.

Reply:
For this, I was thinking of adding like a `maxEntriesPerRow` type setting (sort of like a pagination thing) which would bound the list/map entries for the same grouping key, but the schema would remain `ArrayType/MapType` per row. Was thinking of adding this as a separate change though after the base change was merged

Reply:
I think it's even OK to not have index and just have multiple rows for the same state key. (It's same as explode) Having index would help us to look into detail of how array elements are stored though.

But I don't feel like having 1000s of elements (even not primitive) in ArrayType is great in overall.

Reply:
cc. @cloud-fan I'd like to get some advice from Spark SQL expert. Could you please take a quick look into comments and provide some input? Thanks!

Reply:
Discussed with @cloud-fan . We came up with conclusion that there is no good solution. 

I'm OK with this - we have more time to revisit this (addition of TTL, etc.) and change our mind later if we think there is a better way.

Reply:
Yea sounds good - we'll discuss offline and update the UX as needed